### DISCUSSION TITLE
0017 XLS-17d: XFL Developer-friendly representation of XRPL balances

### DISCUSSION THREAD

**Discussion by RichardAH** [XRPLF/XRPL-Standards#39] [XLS Proposals]
## The earlier version can be found by looking at the edit history.

**Comment by Mwni** [comment]
Is it safe to assume that this new format will have to come with its own dedicated math library?

**Reply by RichardAH** [reply]
XFL is just a representation for the number system already present inside `rippled` / on the XRPL. We can pull the classes out of rippled for a C++ math library, if one is needed. For any other language, yes a new library would be needed if you wish to do exactly equivalent computations on these numbers.

**Reply by Mwni** [reply]
So then using the native `rippled` math kit implies that one always has to convert their XFLs into the `rippled` format before doing operations? Wouldn't a 100% XFL workflow be a thing worth aiming at?

**Comment by Mwni** [comment]
There is a small mismatch with the reference implementation. The line
```
let xfl = (is_negative ? 1n : 0n);
```
should be the other way around

**Comment by Mwni** [comment]
I propose a minor change in the binary representation in order to make XFLs sortable by their casted 64 bit integer value:

- bit 62 now is: 0 = negative, 1 = positive
- bit 61 - 53: if the XFL is negative, XOR with `0xFFFF`
- bit 53 - 0: if the XFL is negative, negate, then XOR with `0x3FFFFFFFFFFFFF` 

This will allow for ultra fast sorting using any int64 capable sorting algorithm and will make the default logical operators work out of the box. A working Javascript implementation of this change can be found in the [xrplkit/xfl library](https://github.com/xrplworks/xrplkit/blob/39905e03ae1909b184dd4bd49cc596051531e80e/xfl/conversion/bigint.js#L42)

**Comment by RichardAH** [comment]
As discussed some months ago with @Mwni my preferred solution to the problem of sort-ability is to introduce packed vs unpacked XFLs. 

Normal XFLs are considered packed, and are not sortable. However unpacking them into a sortable format is very inexpensive:
```c++
int64_t unpack_xfl(int64_t packed)
{
    if (is_negative(packed))
        return ~packed;
    return packed;
}

int64_t pack_xfl(int64_t unpacked)
{
    if (unpacked < 0)
        return ~unpacked;
    return unpacked;
}
```

**Comment by mvadari** [comment]
@RichardAH should this spec be moved into PR and kept in the repo or should it be closed? Note that this does not mean that the spec has been accepted/implemented/finalized.

I will move it to PR by copy-pasting the existing draft if there is no response in the next week.
